### PR TITLE
fast route: avoid repetition

### PR DESCRIPTION
**Pull request by oharboe**

### CODE REVIEW

**Comment by maliberty** [comment]
This is nice but there are other designs that could be similar updated (eg mock-alu).  Would you be willing to do a full update?

The new variable should be documented.

**Comment by oharboe** [comment]
i would limit myself to asap7. reasonable?

**Comment by maliberty** [comment]
> i would limit myself to asap7. reasonable?

I'll take what I can get

**Comment by oharboe** [comment]
@maliberty Removed all FASTROUTE_TCL for asap7, the other PDKs require more care because clocks are also involved...

**Review by maliberty — APPROVED** [review]

**Comment by oharboe** [comment]
@maliberty are these errors on master? familiar?

**Comment by oharboe** [comment]
Trying the below(failed in CI) with `./build_openroad --latest --local`:

```
make DESIGN_CONFIG=designs/nangate45/aes/config.mk metadata
```

Lots of time spent in retiming... TNS_END_PERCENT=100 is futile...

```
[deleted]
[WARNING ODB-0437] Pin directions are ignored from floorplan DEF files.
[WARNING ODB-0437] Pin directions are ignored from floorplan DEF files.
[WARNING ODB-0437] Pin directions are ignored from floorplan DEF files.
[WARNING ODB-0437] Pin directions are ignored from floorplan DEF files.
[WARNING ODB-0437] Pin directions are ignored from floorplan DEF files.
[WARNING ODB-0437] Pin directions are ignored from floorplan DEF files.
[WARNING ODB-0437] Pin directions are ignored from floorplan DEF files.
[WARNING ODB-0437] Pin directions are ignored from floorplan DEF files.
[deleted]
[WARNING RSZ-0075] makeBufferedNet failed for driver sa00_sr\[0\]$_DFF_P_/Q
[WARNING RSZ-0075] makeBufferedNet failed for driver _21201_/ZN
[WARNING RSZ-0075] makeBufferedNet failed for driver _22547_/ZN
[WARNING RSZ-0075] makeBufferedNet failed for driver _23894_/Z
[deleted]
repair_timing -verbose -repair_tns 100
[INFO RSZ-0094] Found 160 endpoints with setup violations.
[INFO RSZ-0099] Repairing 160 out of 160 (100.00%) violating endpoints...
   Iter   | Removed | Resized | Inserted | Cloned |  Pin  |    WNS   |   TNS      |  Viol  | Worst
          | Buffers |  Gates  | Buffers  |  Gates | Swaps |          |            | Endpts | Endpt
---------------------------------------------------------------------------------------------------
        0 |       0 |       0 |        0 |      0 |     0 |   -0.165 |      -16.9 |    160 | sa30_sr\[3\]$_DFF_P_/D
       10 |       0 |       2 |       12 |      0 |     3 |   -0.148 |      -16.6 |    160 | sa20_sr\[3\]$_DFF_P_/D
       20 |       0 |       5 |       27 |      0 |     4 |   -0.143 |      -16.4 |    160 | sa11_sr\[3\]$_DFF_P_/D
       30 |       0 |      11 |       32 |      0 |     6 |   -0.150 |      -16.2 |    160 | sa11_sr\[3\]$_DFF_P_/D
       40 |       0 |      15 |       46 |      0 |     7 |   -0.135 |      -15.6 |    160 | sa20_sr\[3\]$_DFF_P_/D
       50 |       0 |      18 |       63 |      0 |     8 |   -0.132 |      -15.4 |    160 | sa10_sr\[2\]$_DFF_P_/D
       60 |       0 |      22 |       73 |      0 |    11 |   -0.131 |      -15.2 |    160 | sa30_sr\[3\]$_DFF_P_/D
       70 |       1 |      25 |       84 |      0 |    13 |   -0.125 |      -14.9 |    160 | sa11_sr\[3\]$_DFF_P_/D
       80 |       1 |      27 |       95 |      0 |    17 |   -0.121 |      -14.6 |    160 | sa30_sr\[3\]$_DFF_P_/D
       90 |       1 |      33 |      104 |      0 |    18 |   -0.119 |      -14.4 |    160 | u0.u3.d\[1\]$_DFF_P_/D
      100 |       1 |      37 |      113 |      0 |    20 |   -0.115 |      -14.1 |    160 | sa03_sr\[2\]$_DFF_P_/D
      110 |       2 |      42 |      115 |      0 |    23 |   -0.115 |      -13.9 |    160 | sa31_sr\[0\]$_DFF_P_/D
      120 |       3 |      47 |      120 |      0 |    25 |   -0.119 |      -13.7 |    160 | u0.u1.d\[6\]$_DFF_P_/D
      130 |       3 |      50 |      127 |      0 |    29 |   -0.118 |      -13.5 |    160 | sa10_sr\[2\]$_DFF_P_/D
      140 |       3 |      56 |      132 |      0 |    31 |   -0.109 |      -13.4 |    160 | sa02_sr\[0\]$_DFF_P_/D
      150 |       3 |      61 |      144 |      0 |    32 |   -0.107 |      -13.0 |    160 | sa03_sr\[0\]$_DFF_P_/D
      160 |       3 |      65 |      144 |      1 |    37 |   -0.108 |      -12.9 |    160 | sa10_sr\[3\]$_DFF_P_/D
      170 |       3 |      73 |      146 |      1 |    38 |   -0.105 |      -12.9 |    160 | sa20_sr\[7\]$_DFF_P_/D
      180 |       3 |      80 |      148 |      1 |    40 |   -0.104 |      -12.7 |    160 | sa01_sr\[0\]$_DFF_P_/D
      190 |       3 |      84 |      153 |      1 |    44 |   -0.105 |      -12.6 |    160 | sa11_sr\[4\]$_DFF_P_/D
      200 |       3 |      89 |      153 |      1 |    49 |   -0.106 |      -12.6 |    160 | sa11_sr\[4\]$_DFF_P_/D
      208 |       3 |      85 |      155 |      1 |    49 |   -0.103 |      -12.6 |    160 | sa20_sr\[3\]$_DFF_P_/D
      210 |       3 |      86 |      155 |      1 |    49 |   -0.103 |      -12.5 |    160 | sa20_sr\[3\]$_DFF_P_/D
      220 |       4 |      92 |      155 |      1 |    52 |   -0.103 |      -12.4 |    160 | sa20_sr\[3\]$_DFF_P_/D
      230 |       4 |      98 |      158 |      4 |    52 |   -0.105 |      -12.4 |    160 | sa11_sr\[3\]$_DFF_P_/D
      233 |       4 |      93 |      158 |      2 |    53 |   -0.103 |      -12.4 |    160 | sa20_sr\[3\]$_DFF_P_/D
      239 |       4 |      93 |      160 |      5 |    54 |   -0.103 |      -12.4 |    160 | sa20_sr\[3\]$_DFF_P_/D
      240 |       4 |      93 |      160 |      5 |    54 |   -0.103 |      -12.4 |    160 | sa20_sr\[3\]$_DFF_P_/D
      250 |       4 |     101 |      160 |      7 |    54 |   -0.103 |      -12.3 |    160 | sa20_sr\[3\]$_DFF_P_/D
      260 |       4 |     107 |      160 |      8 |    57 |   -0.103 |      -12.3 |    160 | sa20_sr\[3\]$_DFF_P_/D
      264 |       4 |     108 |      160 |      9 |    59 |   -0.103 |      -12.3 |    160 | sa20_sr\[3\]$_DFF_P_/D
      270 |       4 |     111 |      162 |     10 |    59 |   -0.106 |      -12.3 |    160 | sa11_sr\[3\]$_DFF_P_/D
      280 |       4 |     117 |      162 |     10 |    63 |   -0.111 |      -12.3 |    160 | sa11_sr\[3\]$_DFF_P_/D
      290 |       4 |     123 |      162 |     13 |    64 |   -0.108 |      -12.3 |    160 | sa11_sr\[4\]$_DFF_P_/D
      298 |       4 |     112 |      161 |      9 |    66 |   -0.103 |      -12.3 |    160 | sa20_sr\[3\]$_DFF_P_/D
      300 |       4 |     113 |      161 |      9 |    66 |   -0.103 |      -12.2 |    160 | sa20_sr\[3\]$_DFF_P_/D
      310 |       4 |     119 |      161 |      9 |    70 |   -0.103 |      -12.2 |    160 | sa20_sr\[3\]$_DFF_P_/D
      320 |       4 |     128 |      163 |      9 |    70 |   -0.103 |      -12.1 |    160 | sa20_sr\[3\]$_DFF_P_/D
      330 |       4 |     135 |      163 |     10 |    72 |   -0.103 |      -12.1 |    159 | sa20_sr\[3\]$_DFF_P_/D
      330 |       4 |     135 |      163 |     10 |    72 |   -0.103 |      -12.1 |    158 | sa20_sr\[3\]$_DFF_P_/D
      340 |       4 |     139 |      169 |     10 |    75 |   -0.103 |      -12.0 |    158 | sa20_sr\[3\]$_DFF_P_/D
      349 |       4 |     143 |      171 |     11 |    78 |   -0.103 |      -12.0 |    158 | sa20_sr\[3\]$_DFF_P_/D
      350 |       4 |     143 |      171 |     11 |    78 |   -0.103 |      -12.0 |    158 | sa20_sr\[3\]$_DFF_P_/D
      360 |       5 |     148 |      176 |     12 |    79 |   -0.113 |      -12.0 |    158 | sa20_sr\[5\]$_DFF_P_/D
      367 |       5 |     146 |      173 |     11 |    82 |   -0.103 |      -11.9 |    158 | sa02_sr\[5\]$_DFF_P_/D
      370 |       5 |     147 |      176 |     11 |    82 |   -0.103 |      -11.9 |    158 | sa02_sr\[5\]$_DFF_P_/D
      371 |       5 |     147 |      176 |     11 |    82 |   -0.103 |      -11.9 |    158 | sa02_sr\[5\]$_DFF_P_/D
      380 |       5 |     151 |      181 |     12 |    83 |   -0.103 |      -11.7 |    158 | sa02_sr\[5\]$_DFF_P_/D
      390 |       5 |     159 |      183 |     12 |    84 |   -0.103 |      -11.6 |    158 | sa02_sr\[5\]$_DFF_P_/D
      400 |       5 |     167 |      183 |     12 |    86 |   -0.103 |      -11.6 |    158 | sa32_sr\[3\]$_DFF_P_/D
      410 |       5 |     175 |      189 |     12 |    86 |   -0.103 |      -11.5 |    158 | sa32_sr\[3\]$_DFF_P_/D
      420 |       5 |     183 |      192 |     12 |    87 |   -0.103 |      -11.5 |    158 | sa32_sr\[3\]$_DFF_P_/D
      430 |       5 |     191 |      192 |     14 |    87 |   -0.103 |      -11.5 |    158 | sa02_sr\[5\]$_DFF_P_/D
      440 |       5 |     197 |      194 |     16 |    88 |   -0.103 |      -11.5 |    158 | sa02_sr\[5\]$_DFF_P_/D
      450 |       6 |     203 |      194 |     19 |    88 |   -0.103 |      -11.4 |    158 | sa02_sr\[5\]$_DFF_P_/D
      460 |       6 |     210 |      196 |     19 |    90 |   -0.103 |      -11.4 |    158 | sa02_sr\[5\]$_DFF_P_/D
      462 |       6 |     207 |      195 |     19 |    92 |   -0.103 |      -11.4 |    158 | sa02_sr\[5\]$_DFF_P_/D
      470 |       6 |     209 |      203 |     19 |    94 |   -0.103 |      -11.3 |    158 | sa02_sr\[5\]$_DFF_P_/D
      480 |       6 |     213 |      208 |     19 |    98 |   -0.103 |      -11.2 |    158 | sa02_sr\[5\]$_DFF_P_/D
      486 |       6 |     214 |      212 |     19 |    99 |   -0.103 |      -11.1 |    158 | sa02_sr\[5\]$_DFF_P_/D
      490 |       6 |     216 |      214 |     19 |    99 |   -0.103 |      -11.1 |    158 | sa02_sr\[5\]$_DFF_P_/D
      500 |       6 |     219 |      216 |     19 |   105 |   -0.103 |      -11.0 |    158 | sa02_sr\[5\]$_DFF_P_/D
      510 |       7 |     226 |      216 |     19 |   107 |   -0.103 |      -11.0 |    158 | sa02_sr\[5\]$_DFF_P_/D
      515 |       7 |     228 |      216 |     19 |   109 |   -0.103 |      -11.0 |    158 | sa02_sr\[5\]$_DFF_P_/D
      520 |       7 |     230 |      216 |     19 |   111 |   -0.103 |      -11.0 |    158 | sa02_sr\[5\]$_DFF_P_/D
      530 |       7 |     234 |      223 |     20 |   113 |   -0.103 |      -11.0 |    158 | sa02_sr\[5\]$_DFF_P_/D
      540 |       8 |     239 |      223 |     20 |   117 |   -0.103 |      -11.0 |    158 | sa02_sr\[5\]$_DFF_P_/D
      540 |       7 |     232 |      221 |     20 |   117 |   -0.103 |      -10.9 |    158 | sa02_sr\[5\]$_DFF_P_/D
      550 |       7 |     235 |      231 |     20 |   120 |   -0.103 |      -10.8 |    158 | sa02_sr\[5\]$_DFF_P_/D
      560 |       9 |     238 |      235 |     20 |   123 |   -0.103 |      -10.7 |    158 | sa02_sr\[5\]$_DFF_P_/D
      570 |       9 |     240 |      247 |     20 |   126 |   -0.103 |      -10.6 |    158 | sa02_sr\[5\]$_DFF_P_/D
      580 |      10 |     245 |      249 |     21 |   128 |   -0.103 |      -10.6 |    158 | sa02_sr\[5\]$_DFF_P_/D
      584 |       7 |     234 |      239 |     20 |   131 |   -0.103 |      -10.8 |    158 | sa02_sr\[5\]$_DFF_P_/D
      590 |       7 |     237 |      244 |     20 |   131 |   -0.103 |      -10.9 |    158 | sa02_sr\[5\]$_DFF_P_/D
      600 |       7 |     245 |      246 |     20 |   132 |   -0.103 |      -10.8 |    158 | sa02_sr\[5\]$_DFF_P_/D
      610 |       7 |     252 |      248 |     20 |   134 |   -0.103 |      -10.8 |    158 | sa02_sr\[5\]$_DFF_P_/D
      611 |       7 |     249 |      248 |     20 |   135 |   -0.103 |      -10.8 |    158 | sa02_sr\[5\]$_DFF_P_/D
      620 |       7 |     252 |      251 |     20 |   139 |   -0.103 |      -10.8 |    158 | sa02_sr\[5\]$_DFF_P_/D
      624 |       7 |     252 |      251 |     20 |   141 |   -0.103 |      -10.8 |    158 | sa02_sr\[5\]$_DFF_P_/D
      630 |       7 |     255 |      254 |     20 |   142 |   -0.103 |      -10.8 |    158 | sa02_sr\[5\]$_DFF_P_/D
      640 |       7 |     261 |      256 |     20 |   145 |   -0.103 |      -10.8 |    158 | sa02_sr\[5\]$_DFF_P_/D
      649 |       7 |     266 |      256 |     21 |   148 |   -0.103 |      -10.8 |    158 | sa02_sr\[5\]$_DFF_P_/D
      650 |       7 |     266 |      256 |     21 |   148 |   -0.103 |      -10.8 |    158 | sa02_sr\[5\]$_DFF_P_/D
      660 |       7 |     270 |      263 |     21 |   151 |   -0.103 |      -11.0 |    158 | sa02_sr\[5\]$_DFF_P_/D
      670 |       7 |     276 |      263 |     21 |   155 |   -0.103 |      -10.9 |    158 | sa02_sr\[5\]$_DFF_P_/D
      679 |       7 |     282 |      269 |     21 |   156 |   -0.103 |      -10.9 |    158 | sa02_sr\[5\]$_DFF_P_/D
      680 |       7 |     282 |      269 |     21 |   156 |   -0.103 |      -10.9 |    158 | sa02_sr\[5\]$_DFF_P_/D
      690 |       7 |     287 |      276 |     23 |   156 |   -0.109 |      -11.1 |    158 | sa20_sr\[5\]$_DFF_P_/D
      700 |       7 |     296 |      276 |     23 |   157 |   -0.103 |      -11.1 |    158 | sa02_sr\[5\]$_DFF_P_/D
      708 |       7 |     289 |      275 |     22 |   160 |   -0.103 |      -10.9 |    158 | sa02_sr\[5\]$_DFF_P_/D
      710 |       7 |     289 |      275 |     22 |   161 |   -0.103 |      -10.8 |    158 | sa02_sr\[5\]$_DFF_P_/D
      720 |       7 |     294 |      280 |     23 |   163 |   -0.103 |      -10.7 |    158 | sa02_sr\[5\]$_DFF_P_/D
      730 |       7 |     297 |      290 |     24 |   166 |   -0.103 |      -10.7 |    158 | sa02_sr\[5\]$_DFF_P_/D
      737 |       7 |     300 |      292 |     25 |   168 |   -0.103 |      -10.6 |    158 | sa02_sr\[5\]$_DFF_P_/D
      740 |       7 |     300 |      294 |     25 |   169 |   -0.103 |      -10.6 |    158 | sa02_sr\[5\]$_DFF_P_/D
      750 |       7 |     306 |      302 |     25 |   170 |   -0.103 |      -10.5 |    158 | sa02_sr\[5\]$_DFF_P_/D
      760 |       7 |     311 |      310 |     26 |   171 |   -0.103 |      -10.4 |    158 | sa02_sr\[5\]$_DFF_P_/D
      770 |       7 |     317 |      315 |     26 |   173 |   -0.103 |      -10.6 |    158 | sa02_sr\[5\]$_DFF_P_/D
      780 |       7 |     322 |      319 |     26 |   176 |   -0.103 |      -10.6 |    158 | sa02_sr\[5\]$_DFF_P_/D
      790 |       8 |     328 |      322 |     26 |   178 |   -0.103 |      -10.6 |    158 | sa02_sr\[5\]$_DFF_P_/D
      797 |       8 |     326 |      322 |     26 |   181 |   -0.103 |      -10.5 |    158 | sa02_sr\[5\]$_DFF_P_/D
      800 |       8 |     326 |      326 |     26 |   181 |   -0.103 |      -10.5 |    158 | sa02_sr\[5\]$_DFF_P_/D
      810 |       8 |     327 |      341 |     26 |   185 |   -0.103 |      -10.4 |    158 | sa02_sr\[5\]$_DFF_P_/D
      814 |       8 |     327 |      340 |     26 |   186 |   -0.103 |      -10.4 |    158 | sa02_sr\[5\]$_DFF_P_/D
      820 |       8 |     331 |      340 |     26 |   187 |   -0.103 |      -10.3 |    158 | sa02_sr\[5\]$_DFF_P_/D
      830 |       8 |     336 |      343 |     27 |   190 |   -0.103 |      -10.3 |    158 | sa02_sr\[5\]$_DFF_P_/D
      836 |       8 |     341 |      343 |     27 |   191 |   -0.103 |      -10.3 |    158 | sa02_sr\[5\]$_DFF_P_/D
      840 |       8 |     342 |      345 |     27 |   192 |   -0.103 |      -10.2 |    158 | sa02_sr\[5\]$_DFF_P_/D
      850 |       8 |     346 |      356 |     27 |   194 |   -0.103 |      -10.2 |    158 | sa02_sr\[5\]$_DFF_P_/D
      860 |      10 |     351 |      360 |     27 |   195 |   -0.103 |      -10.1 |    158 | sa02_sr\[5\]$_DFF_P_/D
      866 |       8 |     350 |      359 |     27 |   196 |   -0.103 |      -10.1 |    158 | sa02_sr\[5\]$_DFF_P_/D
      870 |       8 |     350 |      359 |     28 |   198 |   -0.103 |      -10.1 |    158 | sa02_sr\[5\]$_DFF_P_/D
```

File: flow/scripts/util.tcl [hunk 1]
```diff
@@ -2,3 +2,15 @@ proc log_cmd {cmd args} {
   puts "$cmd [join $args " "]"
   $cmd {*}$args
 }
+
+proc fast_route {} {
+  if {[info exist env(FASTROUTE_TCL)]} {
```
Comment (maliberty) [inline flow/scripts/util.tcl]:
```suggestion
  if {[info exist ::env(FASTROUTE_TCL)]} {
```

**Comment by maliberty** [comment]
master is passing.

**Comment by oharboe** [comment]
@maliberty all green PR now latest fixes

**Pull request merged**